### PR TITLE
Load test findings

### PR DESCRIPTION
**Pull request by RhysJones125**

### CODE REVIEW

File: load_testing.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# Assumptions
+
+We have used a chrome plug-in, SimilarWeb to investigate other sites which could have a similar audience as our dashboards - LG Inform and LGC plus. These sites seem to average in the range of 50k users a month.
+
+The load test will initally focus on user routes through the dashboard, instead of simply requesting each page indidually then moving onto the next test. The benefit of this approach is we can mimic a user making choices in the available dropdowns on each of the pages.
+
+The first iteration of load testing will not use a web framework such as Cypress or Selenium WebDriver to control changes to the dropdowns on each of the pages. Instead the locust tests will issue new page requests with query strings values matching the desired options in the dropdowns. This will slightly skew the results as instead of an asynchronous XHR request being executed within the page, we are requesting a new page load that will have a larger page size.
+
+The metrics for these load tests will be gathered using the Prometheus platform logging for CPU usage and the Locust user interface will provide response times and failed user requests.
```
Comment (chopkinsmade) [inline load_testing.md:9]: Prometheus will be used to gather the CPU Usage and Request Per Second metrics; the Locust user interface will provide response time and failed user request metrics

Maybe something like this instead?

File: load_testing.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+![total_requests_per_second_general](images/load_test/general/total_requests_per_second_1651672444.png) <br>
+There were roughly 28 requests per second, of which none failed.
+
+## Launch day
+We ran a load test where we spawned 10 users per second up to a maximum of 460 users, we ran this test for 15 minutes.
+We repeated this test 3 times and the results were always within the same bounds. 
+
+![CPU_launch](images/load_test/launch/cpu_launch_usage.PNG) <br>
+CPU usage was consistently around 96% for most of the load test, however at some points it exceeded 100% and reached 110%.
```
Comment (robertblincoe) [inline load_testing.md:43]: Good clear explanation of how you estimated usage and tested it

File: load_testing.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+![response_times_launch](images/load_test/launch/response_times_(ms)_1651673914.png) <br>
+The 95 percentile response time averaged at 400ms, however the average median time was around 80ms. There was more fluctuation for the response time with the higher number of users, which is to be expected.
+
+![total_requests_per_second_launch](images/load_test/launch/total_requests_per_second_1651673914.png) <br>
+There were roughly 200 requests per second, of which none failed.
+
+# Conclusion
+Based on our estimated usage of the dashboard, we expect it to be able to handle loads of up to 460 users at one time. However, at higher usage we can expect poorer performance, though this should not be longer than 1 second, which we believe to be within a reasonable range.
```
Comment (robertblincoe) [inline load_testing.md:52]: Is there any Government guidance about response rates and what are the acceptable limits? 1s at worst sounds OK to me.

**Review by robertblincoe — APPROVED** [review]

File: load_testing.md (hunk 3 above)
Comment (RhysJones125) [reply load_testing.md:52]: I'm not sure of any government standard, but there is stack overflow comment which is pretty useful. https://stackoverflow.com/questions/164175/what-is-considered-a-good-response-time-for-a-dynamic-personalized-web-applicat

File: load_testing.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
+# Assumptions
+
+We have used a chrome plug-in, SimilarWeb to investigate other sites which could have a similar audience as our dashboards - LG Inform and LGC plus. These sites seem to average in the range of 50k users a month.
+
+The load test will initally focus on user routes through the dashboard, instead of simply requesting each page indidually then moving onto the next test. The benefit of this approach is we can mimic a user making choices in the available dropdowns on each of the pages.
+
+The first iteration of load testing will not use a web framework such as Cypress or Selenium WebDriver to control changes to the dropdowns on each of the pages. Instead the locust tests will issue new page requests with query strings values matching the desired options in the dropdowns. This will slightly skew the results as instead of an asynchronous XHR request being executed within the page, we are requesting a new page load that will have a larger page size.
```
Comment (adrianclay) [inline load_testing.md:7]: Hmm, assumption here that the XHR is smaller than the page load.  I'm not sure that is necessarily correct?

See screenshot where the XHR request is 1.5Mb and the slowest response.
image

File: load_testing.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+Breaking that down into a 20 minute window equates to around 46 users.
+
+
+## Launch day estimate:
+We estimate the number of users on launch day to be 10 times that of our general usage estimate, equating to 460 users.
+
+
+# Findings
+## General usage
+We ran a load test where we spawned 1 user per second up to a maximum of 46 users, we ran this test for 15 minutes.
```
Comment (adrianclay) [inline load_testing.md:25]: Given the above comment, could I see the code/reproduce this load test?

File: load_testing.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
+# Assumptions
+
+We have used a chrome plug-in, SimilarWeb to investigate other sites which could have a similar audience as our dashboards - LG Inform and LGC plus. These sites seem to average in the range of 50k users a month.
+
+The load test will initally focus on user routes through the dashboard, instead of simply requesting each page indidually then moving onto the next test. The benefit of this approach is we can mimic a user making choices in the available dropdowns on each of the pages.
+
+The first iteration of load testing will not use a web framework such as Cypress or Selenium WebDriver to control changes to the dropdowns on each of the pages. Instead the locust tests will issue new page requests with query strings values matching the desired options in the dropdowns. This will slightly skew the results as instead of an asynchronous XHR request being executed within the page, we are requesting a new page load that will have a larger page size.
+
+Prometheus will be used to gather the CPU Usage and Request Per Second metrics; the Locust user interface will provide response time and failed user request metrics
```
Comment (adrianclay) [inline load_testing.md:9]: Did Prometheus give any metrics on Memory usage?

**Review by adrianclay — COMMENTED** [review]
Some discussion points worth bottoming out, however I don't think my comments affect the conclusion.

More general comment: This `load_testing.md` document is both specific to Local Government users, but could also be useful for other dashboards, so while I find it odd that it's in this repo I don't have a good alternative.

File: load_testing.md (hunk 5 above)
Comment (kateriley1) [reply load_testing.md:25]: Instructions on how to run a load test locally can be found in the LG repo [here](https://github.com/communitiesuk/local_government_dashboard/blob/main/README.md#load-tests)

File: load_testing.md (hunk 6 above)
Comment (kateriley1) [reply load_testing.md]: We have now included memory utilisation graphs for the general and launch day load tests.

File: load_testing.md (hunk 5 above)
Comment (kateriley1) [reply load_testing.md:25]: The code used to run the load test can be found [here](https://github.com/communitiesuk/local_government_dashboard/blob/main/tests/load_test/locustfile.py)

**Review by backslashbaker — APPROVED** [review]

File: load_testing.md (hunk 4 above)
Comment (RhysJones125) [reply load_testing.md]: We've updated that assumption and included a section on where we go from this to improve this load test

**Pull request merged**